src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs

```diff
@@ -2,81 +2,108 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
+using System;
 using System.Collections;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Serialization
+namespace Microsoft.CodeAnalysis.Serialization;
+
+/// <summary>
+/// collection which children is checksum.
+/// </summary>
+internal sealed class ChecksumCollection(ImmutableArray<Checksum> children) : IReadOnlyCollection<Checksum>, IChecksummedObject
```

Comment:
majorly simplified the concepts here.  There are now only two types of interest, and no inheritance.  there is simply:

Checksum.  A raw checksum with no additional info.
ChecksumCollection.  A collection of checksums, and an aggregated checksum aggregating all of the child checks.

ChecksumCollection only contains checksums (not other ChecksumCollections).

Finally, there is SolutionChecksums/ProjectChecksums/DocumentChecksums.  These are just simple types, not in the checksum hierarchy that only contain Checksum/ChecksumCollections internally.

Reply:
i would like to get rid of this `IReadOnlyCollection<Checksum>` inheritance as well.  but we have some codepaths that uniformly process `HashSet<Checksum>` and this type.  I intend to probably split those as well.